### ISSUE TITLE
Add support for component private pipes or pure method calls from templates

### ISSUE THREAD

**Issue by robmv** [angular/angular#47534]
### Description

Currently pipes must be declared at module level. Sometimes a pipe is only used on a single component. The power of pipes being by default pure is that they are only invoked when the input changes. Using a local component function force Angular to always run the function because it has no way to know if the function is pure or not.

### Proposed solution

Add a `declarations` setting to the @Component decorator that allows declaring pipes locally for the component. Another option could be to add a new @Pure decorator that can be added to the component method in order to tell Angular that the function invocation from the template gives pure results.

### Alternatives considered

Declaring pipes on the module level, this still add too much boilerplate when the pipe is used on a single place.

**Comment by pkozlowski-opensource** [comment]
> Add a declarations setting to the Component decorator that allows declaring pipes locally for the component

You can achieve this today with standalone components.

**Comment by robmv** [comment]
> You can achieve this today with standalone components.

This is great, I thought that was only for components with a view and not pipelines. I wonder what is the overhead of having a lot of standalone pipelines instead of Angular restricting a non standalone pipeline to be used inside a component? Do a standalone pipeline used on a single component add more overhead than a pipeline declared on a module?

**Comment by alxhub** [comment]
Nope, there is no additional overhead for a pipe that's only used on one component - in fact it will probably be less overhead.

**Comment by pkozlowski-opensource** [comment]
As noted, standalone components support locally-defined pipes, ex.:

```typescript
@Pipe({ name: 'brackets', standalone: true })
export class ExamplePipe implements PipeTransform {
  transform(value: any, ...args: any[]) {
    return `(${value})`;
  }
}

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [ExamplePipe],
  template: `{{'I am in brackets' | brackets}}`,
})
export class ExampleStandaloneComponent {}
```

Working stackblitz: https://stackblitz.com/edit/angular-standalone-ks9pkt?file=src%2Fmain.ts

There is no performance penalty for using this approach.

Closing as already implemented.
